File: sonarqube-companion-frontend/src/app/violations/group-violations-history-diff.ts
import {Violations} from './violations';
import {ProjectViolationsHistoryDiff} from './project-violations-history-diff';

export class GroupViolationsHistoryDiff {

  groupDiff: Violations;
  addedViolations: Violations;
  removedViolations: Violations;
  projects: {[key: string]: ProjectViolationsHistoryDiff} = {};

  constructor(data) {
    this.groupDiff = new Violations(data.groupDiff);
    this.addedViolations = new Violations(data.addedViolations);
    this.removedViolations = new Violations(data.removedViolations);
    data.projectDiffs.forEach(projectDiff => {
      this.projects[projectDiff.projectKey] = new ProjectViolationsHistoryDiff(projectDiff);
    });
  }

}

File: sonarqube-companion-rest/src/main/java/pl/consdata/ico/sqcompanion/violation/project/GroupViolationsHistoryDiff.java
package pl.consdata.ico.sqcompanion.violation.project;

import lombok.*;
import pl.consdata.ico.sqcompanion.violation.Violations;

import java.util.ArrayList;
import java.util.List;

@Data
@Builder
@NoArgsConstructor
@AllArgsConstructor
public class GroupViolationsHistoryDiff {

    private Violations groupDiff;
    private Violations addedViolations;
    private Violations removedViolations;
    private List<ProjectViolationsHistoryDiff> projectDiffs;

    public static GroupViolationsHistoryDiff empty() {
        return GroupViolationsHistoryDiff.builder()
                .addedViolations(Violations.builder()
                        .blockers(0)
                        .criticals(0)
                        .majors(0)
                        .minors(0)
                        .infos(0)
                        .build())
                .removedViolations(Violations.builder()
                        .blockers(0)
                        .criticals(0)
                        .majors(0)
                        .minors(0)
                        .infos(0)
                        .build())
                .groupDiff(Violations.builder()
                        .blockers(0)
                        .criticals(0)
                        .majors(0)
                        .minors(0)
                        .infos(0)
                        .build())
                .projectDiffs(new ArrayList<>())
                .build();
    }
}
